worksheet_D/oxo.py

```diff
@@ -0,0 +1,110 @@
+class OxoBoard:
+    def __init__(self):
+        self.board = {(0, 0): 0,(1, 0): 0, (2, 0): 0,
+                      (0, 1): 0, (1, 1): 0, (2, 1): 0,
+                      (0, 2): 0, (1, 2): 0, (2, 2): 0}      # Creates dictionary.
+
+    def get_square(self, x, y):
+        return self.board[x, y]                             # Looks up value with the key (x, y) and returns it
+
+    def set_square(self, x, y, mark):
+        if self.board[x,y] == 0:
+            self.board[x, y] = mark                         # Changes the value assigned to the key to 'mark'
+            return True
+        else:
+            return False
+
+    def is_board_full(self):
+        squares_searched = 0
+        for i in xrange(0, 3):
+            for j in xrange(0, 3):
+                if self.board[i, j] > 0:
+                    squares_searched += 1                   # A counter for number of squares searched
+                if squares_searched == 9:
+                    return True
+        return False
+
+    def get_winner(self):
+
+        if self.board[1, 1] != 0:
+            # checks diagonal
+            if self.board[0, 0] == self.board[1, 1] and self.board[2, 2] == self.board[1, 1]:
```

Comment:
Note that Python allows you to chain `==` together, so e.g. you could rewrite this as
```
if self.board[0, 0] == self.board[1, 1] == self.board[2, 2]:
```